fixtures: register finalizers with all previous fixtures in the stack (take 3)

SalmonMode opened the pull request:
See #6551 and #6438 for context.

Copied from the original PR:

Turns out the root cause was because fixtures were only considering the fixtures requested in their argnames when trying to find out where to register their finalizers at execute time. They weren't considering autouse fixtures, so when this was combined with parameterization, this resulted in fixtures not being torn down when a parameterized, autouse fixture was being torn down to go from one param set to the next.

This also resulted in fixtures being torn down out of order in extreme cases. You can see an example of this in the test that I added (if you run it without this fix).

The solution ultimately required me to have fixtures register their finalizers with all fixtures that came before them in the stack. This list of fixtures came from using request.fixturenames in combination with the parent requests (when relevant) and the fixture's requested argnames to determine a good slicing point. Since this logic could be summed up on its own, I moved it into its own private method with a beefy docstring to explain it.

bluetech commented:
Hi @SalmonMode, this (re)fixes #6436, right?

Also, since following the context is a bit difficult, can you describe what happened to the first two takes, and what is different in this one?

SalmonMode commented:
Correct

Sure! In the first take, I solved an old bug, but, because I used sets for something (not realizing that order wasn't retained in a set), I introduced another bug, so it had to be reverted. In take 2, @blueyed remade my branch, but we had it use normal lists ones and we added a test to cover it. Unfortunately, @blueyed couldn't dedicate time to finishing this effort, so closed the take 2 PR. I've now remade it again for a third try.

nicoddemus commented:
(Sorry for the delay, this as you know is a sensitive PR so it requires me to reserve a good chunk of free time to evaluate it)

RonnyPfannschmidt commented:
Just as FYI, for me it may take until mid October before I can commit the time for a deep review of this for myself, so feel free to get it done without me

SalmonMode commented on testing/python/fixtures.py:
```diff
@@ -1748,6 +1748,134 @@ def test_world(self):
+            "        SETUP    F f_fix['f2']\n"
+            "        test_auto.py::TestFixtures::test_b[p1-m1-f2] {0}\n"
+            "        TEARDOWN F f_fix['f2']\n"
+            "      TEARDOWN C c_fix\n"
```
The fix basically makes it so that the fixture order is a stack, and a fixture cannot be torn down unless all the fixtures that were setup after it have also been torn down. In this case, I believe `m_fix` is trying to tear down to go from `m1` to `m2`, but to do that, both `c_fix` and `another_c_fix` need to be torn down first. 

This way, there's no surprises related to side effects.

Without this (if `c_fix` would otherwise tear down only at the end of the session), if `c_fix` was unwittingly affected by something `m_fix` did when its param was `m1`, then when `m2` comes around, if `c_fix` wasn't torn down, it would still be influenced by `m1`, while `m_fix` kept doing doings that `c_fix` would no longer be affected by. That would mean that if we ran that class directly with pytest, `test_auto.py::TestFixtures::test_a[p1-m2-f1]` would potentially be doing something different than had we ran `test_auto.py::TestFixtures::test_a[p1-m2-f1]` directly.

For example, if `m_fix` was parameterized so that it was setting up different versioned installations of chrome and chromedriver, and `c_fix` was launching a chrome driver session (not that I recommend handling drivers this way haha), `c_fix` would be executed one and stay as the first driver/browser version installed and driver session made for both `test_auto.py::TestFixtures::test_a[p1-m1-f1]` and `test_auto.py::TestFixtures::test_a[p1-m2-f1]`, while `m_fix` was reinstalling different versions that aren't used by anything. And if `test_auto.py::TestFixtures::test_a[p1-m2-f1]` was run directly, `c_fix` would be executed while a different version of chrome and chromedriver were installed. I'm guessing running the class with pytest (rather than the function directly) would actually throw an error since the driver executable would still be running when `m_fix` tried to uninstall it, but that's the general idea. This fix prevents that sort of thing from happening.

bluetech replied:
This comment nailed the change for me. I agree with your argument, that the new behavior is the correct one. This is indeed a subtle but significant change.

nicoddemus replied:
Thanks a lot for the explanation @SalmonMode, and sorry about the delay, it helped understand the actual behavior behind the change! 

Again, really appreciate the patience on this, given that this is taking months now to review, but is a very sensitive and complex part of the codebase.

The remark about it being now a stack helps a lot.

> if c_fix was unwittingly affected by something m_fix did when its param was m1

I understand the picture you are painting here, the problem I see here is "unwittingly", which is a synonym for implicitly... if `c_fix` is affected by `m_fix`, then `c_fix` should explicitly depend on `m_fix`, I think. 

When I see this code:

```python
@pytest.fixture(scope="class")
def c_fix():
    yield

class TestFixtures:

    def test_a(self, c_fix): pass
    def test_b(self, c_fix): pass
```

In my understanding on how things should work, `c_fix` should only be torn down after all tests in `TestFixtures` have finished, regardless of what other autouse fixtures are in effect in this session, because `c_fix` doesn't depend on anything else.

If `c_fix` has an implicit dependency on `m_fix`, it should explicitly declare that dependency.

The same argument can also be made that this is a bug: if `c_fix` does something costly but completely unrelated to any other fixture, why is it being torn down just because `m_fix` needs to be recreated with a different argument? I understand the new behavior is safer, but also I believe someone might point out this is a regression for the use case above.

A stripped down example:

```python
import pytest

@pytest.fixture(scope="session", params=[1, 2], autouse=True)
def s():
    pass


@pytest.fixture(scope="package", autouse=True)
def p():
    pass

def test():
    pass
```

Given that `p` doesn't depend on anything, I expect it to be setup/torndown exactly once in this session, and this is the behavior on `master`. If I change `p()` to `p(s)` then it is correctly setup and torn down twice, to account for `s[1]` and `s[2]`.

The change here is implicitly adding a dependency to lower scoped fixtures to all their upper scoped fixtures, which I'm not sure it is correct/desirable.

What do you folks think?

Also I need to mention the sheer quality of the PR is remarkable! The documentation improvements are an amazing addition by themselves. 😍

nicoddemus replied:
Hi folks, 

Gentle ping here in case this got lost through the cracks. 👍

SalmonMode replied:
Thanks! It means a lot to hear that you liked my doc changes 😁 

And I absolutely understand how important a change like this is. So I'm _more_ than happy to stick with this as long as it takes.

My perspective is that, in your example, by setting `s` to have a `session` scope, with `autouse`, the goal would be to have it affect everything it can within its reach. Since it's also parameterized, I would expect the goal to be to have each of those param sets be used to affect everything it can within its reach, as well. But if `p` isn't executed once for each param set, then `s` will have failed in its goal to affect everything in its reach.

_That said_, I could understand a desire for the alternate behavior you mentioned. But if that were desired, then, in my opinion, it should be up to the one writing the tests to make those relationships/goals more clear. This could be done by using more appropriate scopes, explicitly requesting fixtures, and/or by relying on structures like classes, modules, nesting packages with multiple `conftest.py` files, and/or grouping tests based on dependencies.

bluetech replied:
@SalmonMode @nicoddemus I wonder what you think of this case:

```py
import pytest

@pytest.fixture(scope="module")
def m():
    print('M SETUP')
    yield
    print('M TEARDOWN')

class TestIt:
    @classmethod
    @pytest.fixture(scope="class", autouse=True)
    def c(cls):
        print('C SETUP')
        yield
        print('C TEARDOWN')

    def test_f1(self):
        print('F1 CALL')

    def test_f2(self, m):
        print('F2 CALL')
```

What should it print?

<details>
<summary>Reveal answer</summary>

In both master and this branch this prints:

```py
C SETUP
F1 CALL
M SETUP
F2 CALL
C TEARDOWN
M TEARDOWN
```

This is not a stack, which seems wrong to me. E.g. in pytest-django I want to use fixtures like this:

```py
@pytest.fixture(scope="...", autouse=...)
def my_fixture():
    db_transaction.begin()
    # set up some DB data here
    yield
    db_transaction.rollback()
```

and this would wreck it!
</details>

SalmonMode replied:
My expectation would be:

```
M SETUP
C SETUP
F1 CALL
F2 CALL
C TEARDOWN
M TEARDOWN
```

Awesome catch! And I'll definitely add this as a test case.

My opinion, is that, at least for the context of the `autouse` fixture (in this case, all the tests inside that class), all fixtures "above" the `autouse` fixture that are relevant to any of the tests inside that context should be automatically relevant to all tests in that context. I believe pytest should be aggressive (for lack of a better word) in its coherence.

I'll find a way to fix this issue.

nicoddemus replied:
Definitely that the module scoped fixture should be created first. 

Off the topic for this PR, but relevant to the subject: time and time again https://github.com/pytest-dev/pytest/issues/4871 comes back to mind and the idea that building and following a proper topology of fixture dependencies would solve this and many other problems, and make things simpler to boot. 😕

nicoddemus replied:
Hey @SalmonMode, sorry for not answering earlier, I planned to answer Yesterday but decided to sleep on it again.

> My perspective is that, in your example, by setting s to have a session scope, with autouse, the goal would be to have it affect everything it can within its reach.

I think you nailed the heart of the issue: should autouse fixtures be applied automatically to other *fixtures* as well?

Currently we have this in the docs:

> The class-level ``transact`` fixture is marked with *autouse=true*
which implies that all test methods in the class will use this fixture
without a need to state it in the test function signature or with a
class-level ``usefixtures`` decorator.

And:

> Here is how autouse fixtures work in other scopes:
> * autouse fixtures obey the scope= keyword-argument: if an autouse fixture has scope='session' it will only be run once, no matter where it is defined. scope='class' means it will be run once per class, etc.
> * if an autouse fixture is defined in a test module, all its test functions automatically use it.
> * if an autouse fixture is defined in a conftest.py file then all tests in all test modules below its directory will invoke the fixture.
> * lastly, and please use that with care: if you define an autouse fixture in a plugin, it will be invoked for all tests in all projects where the plugin is installed. This can be useful if a fixture only anyway works in the presence of certain settings e. g. in the ini-file. Such a global fixture should always quickly determine if it should do any work and avoid otherwise expensive imports or computation.

(From https://github.com/pytest-dev/pytest/blob/master/doc/en/fixture.rst#autouse-fixtures-xunit-setup-on-steroids)

So the intent has always been for autouse fixtures to be implied in test functions, as if the user has explicitly added the fixture name as parameter, but not necessarily for the same to happen to other fixtures within its reach.

I'm not quoting the docs as some "holy truth" that should be followed no matter what, but just to demonstrate that I believe this is how users in general think of an `autouse` fixture: pytest just automatically add the fixture name as parameter to the test function.

Back to the example I gave before:

```python
import pytest

@pytest.fixture(scope="session", params=[1, 2], autouse=True)
def s():
    pass

@pytest.fixture(scope="module", autouse=True)
def p():
    pass

def test():
    pass
```

On `master` we have:

```
SETUP    S s[1]
    SETUP    M p
        test-auto-fix.py::test[1] (fixtures used: p, s).
TEARDOWN S s[1]
SETUP    S s[2]
        test-auto-fix.py::test[2] (fixtures used: p, s).
    TEARDOWN M p
TEARDOWN S s[2]
```

(`p` is created only once).

Here we have:

```
SETUP    S s[1]
    SETUP    M p
        test-auto-fix.py::test[1] (fixtures used: p, s).
    TEARDOWN M p
TEARDOWN S s[1]
SETUP    S s[2]
    SETUP    M p
        test-auto-fix.py::test[2] (fixtures used: p, s).
    TEARDOWN M p
TEARDOWN S s[2]
```

(`p` is created twice).

I suspect this will break tons of code, and unfortunately I don't think there's a simple workaround for code bases that desire the behavior we have on `master` today, while getting the second behavior is easy: all one needs to do is make `p` depend on `s`.

To summarize, my thinking is:

1. Autouse fixtures automatically affect test functions, not fixtures.
2. This change if applied would break existing suites, without a suitable workaround.

Again, utmost respect for your work, and thanks a lot for bearing with us over this PR so long.

SalmonMode replied:
No worries! This is definitely a serious change that can't be taken lightly.

I wouldn't consider this change as "breaking" existing test suites, as much as it is exposing likely risky oversights in fixture dependencies (I noticed this in `pytest-django` after the first take on this change). The current behavior in `master` allows for very inconsistent test execution in that running a test independently, versus running it as part of the entire suite, can have very different results.

I would also argue that fixtures are part of the tests, as they cannot exist without the tests. Currently in `master`, `autouse` does affect automatically affect fixtures, in that it makes it possible for different fixtures to run for a given test, depending on the tests you tell pytest to run.

In the example you provided here:

```
SETUP    S s[1]
    SETUP    M p
        test-auto-fix.py::test[1] (fixtures used: p, s).
TEARDOWN S s[1]
SETUP    S s[2]
        test-auto-fix.py::test[2] (fixtures used: p, s).
    TEARDOWN M p
TEARDOWN S s[2]
```

it demonstrates the sort of coherency issues that I'm referring to (no offense intended here. I mean this in the literal sense).

This example implies that `p` is dependent on `s[1]`. The first test is dependent on `p` and `s[1]`. The second test is dependent on `p`, which is dependent on `s[1]`, but the second test is also dependent on `s[2]`. This means that the second test is dependent on both `s[1]` (via the transitive property) and `s[2]`, and so `s` shouldn't have been parameterized, and it would have been simpler to make it into another fixture with another test for the different test scenario, like this:

```py
import pytest

@pytest.fixture(scope="session", autouse=True)
def s1():
    pass

@pytest.fixture(scope="module", autouse=True)
def p():
    pass

@pytest.fixture
def s2():
    pass

def test1():
    pass

def test2(s2):
    pass
```

yielding this:

```
SETUP    S s1
    SETUP    M p
        test-auto-fix.py::test1 (fixtures used: p, s1)
        SETUP    F s2
            test-auto-fix.py::test2 (fixtures used: p, s1, s2)
        TEARDOWN F s2
    TEARDOWN M p
TEARDOWN S s1
```

Either that, or `p` didn't need to happen after `s`, and it could have been this instead:

```py
import pytest

@pytest.fixture(scope="session", autouse=True)
def p():
    pass

@pytest.fixture(scope="module", params=[1, 2], autouse=True)
def s():
    pass

def test():
    pass
```

resulting in this:

```
SETUP    S p
    SETUP    M s[1]
        test-auto-fix.py::test[1] (fixtures used: p, s).
    TEARDOWN M s[1]
    SETUP    M s[2]
        test-auto-fix.py::test[2] (fixtures used: p, s).
    TEARDOWN M s[2]
TEARDOWN S p
```

_Or_, there's no apparent way to correctly run that second test.

You're right in that there isn't an easy workaround to continue to achieve something like:

```
SETUP    S s[1]
    SETUP    M p
        test-auto-fix.py::test[1] (fixtures used: p, s).
TEARDOWN S s[1]
SETUP    S s[2]
        test-auto-fix.py::test[2] (fixtures used: p, s).
    TEARDOWN M p
TEARDOWN S s[2]
```

after this change, but I would argue that that's because logically incoherent things are inherently complex.

I believe that if people aren't consciously intending to depend on this (and I would guess this the vast majority of those who depend on it), then this change will be a good thing for them, as it will expose these issues in their test suites (like with what I saw in `pytest-django`) so they can be addressed.

For those that _are_ consciously depending on this, I'm guessing they're already trying to write tests in a more complicated way than is necessary, likely due to low internal software quality (short of that one guy who was asking a while back about procedurally generating test cases for an extremely niche testing method regarding chip fabrication)

nicoddemus replied:
> I would also argue that fixtures are part of the tests, as they cannot exist without the tests.

That's a bit vague, but let me word that differently: *autouse* fixtures are automatically applied to *functions*, not to other fixtures. When I mentioned that I was talking specifically about autouse fixtures, which I belive is the main point we are discussing.

> it demonstrates the sort of coherency issues that I'm thinking about (no offense intended here. I mean this in the literal sense).

Sorry, I don't see any inconsistency. :confused:

In that example, pytest needs to execute `test()`. In the same scope we have two autouse fixtures, `s` and `p`, so the test is considered to have been written as if the user explicitly wrote `test(s, p)`. Given that `s` is parametrized, pytest executes two tests, `test(s[1], p)` and `test(s[2], p)`. Because `p` doesn't depend on anything else, there's no reason for pytest to recreate `p`.

> This example implies that p is dependent on s[1].

I disagree... `p` doesn't really depend on anything. Just because a fixture was created before it, doesn't imply a dependency IMHO.

Thanks for the examples, but all the output shown after them (with setup/teardown order of the fixtures involved) makes sense to me, so it is hard for me to understand what is the argument behind them, sorry.

---

I believe the main point of disagreement between us is that you believe that fixtures should explicitly depend on any autouse fixture acting on it, and I don't think they should.

To exemplify:

```python
import pytest

@pytest.fixture(scope="session", params=[1, 2], autouse=True)
def s():
    pass

@pytest.fixture(scope="module")
def m():
    pass

def test(m):
    pass    
```

What happens today, is that pytest executes as _if_ the code was written like this:

```python
import pytest

@pytest.fixture(scope="session", params=[1, 2])
def s():
    pass

@pytest.fixture(scope="module")
def m():
    pass

def test(s, m):  # <--- s added because s is autouse
    pass    
```

What you argue is that pytest _should_ be doing this:

```python
import pytest

@pytest.fixture(scope="session", params=[1, 2])
def s():
    pass

@pytest.fixture(scope="module")
def m(s):  # <--- s added because s is autouse
    pass

def test(s, m):  # <--- s added because s is autouse
    pass    
```

Is my undertanding correct? Please correct me if I'm wrong. 

If my understanding is right, then I argue that the current behavior is correct, and it will definitely break existing suites that are working, and working not by accident, but by design (I say that from our codebase at work).

It is common for an autouse fixture to execute something which is system-wide, but not necessarily affects other fixtures in the system.

I will give an example derived from real code from work:

```python
@pytest.fixture(scope="session", params=["br", "en"], autouse=True)
def set_system_locale(request):
    # sets current system locale to request.param
    ...


# in another module
@pytest.fixture(scope="package", autouse=True)
def initialize_petsc():
    # initializes petsc C++ runtime
    ...
```

In the current behavior, `set_system_locale` doesn't affect `initialize_petsc` (and it should not, petsc doesn't care about the locale at all). We tests with `set_system_locale[br]` and `set_system_locale[en]`, and `initialize_petsc` is executed once. Under your proposal, `initialize_pestc` would execute twice, which is wrong and probably would crash the system.


> I believe that if people aren't consciously intending to depend on this (and I would guess this the vast majority of those who are depending on it), then this change will be a good thing for them, as it will expose these issues in their test suites (like with what I saw in pytest-django) so they can be addressed.

I'm not familiar with the issue with pytest-django you mention... perhaps you can exemplify it to clarify the point better?

Again sorry for all the discussion; I'm not trying to be difficult, I'm really making an effort to make sure I'm not missing the point, but if I'm understanding the proposal correctly, I don't think it is correct.

Perhaps a new example of how things work now, and how you belive they should work, but with "real" examples, might help me understand what you believe the problem is (in case I'm still missing your point, as I tried to exemplify above)? It might be I'm missing something due to the fact that we are using abstract examples (like `s`, `p` and so on)? In that case the pytest-django case might help.

Again thanks for the patience!

SalmonMode replied:
To clarify, I'm not saying that the autouse fixture _has_ to be dependent on the higher level fixture. I'm saying that if there's a need for (in the example), `s[1]` to happen before `p`, then `p` must be dependent on `s[1]`, but if it isn't, then it can happen after `p`, and if it can happen after `p`, then it can be restructured slightly so that `s` happens after `p` and each param for `s` could apply to a separate version of that test without `p` having to be run again. This is where my second example output came in, i.e.:

```
SETUP    S p
    SETUP    M s[1]
        test-auto-fix.py::test[1] (fixtures used: p, s).
    TEARDOWN M s[1]
    SETUP    M s[2]
        test-auto-fix.py::test[2] (fixtures used: p, s).
    TEARDOWN M s[2]
TEARDOWN S p
```

The goal with my change is to get fixtures working like a stack so it's easier to manage/debug fixtures. I'm not saying there is always a dependent implied, but if the setup/teardown plan is jumping around, that suggests to me that things are being tested in contexts that have more going on than is necessary, or because scopes are being leveraged in a certain way to control ordering of behavior (and I believe that way could be made more effective/convenient/easy to read).

In regards to your work example, things get tricky, because I don't know where that first fixture is defined. If it's in your top-level contest, for example, then I would expect that second fixture to run twice, and if that's not desired, then I wouldn't have that first fixture in the top level conftest. 

That said, I've been thinking about this for a while because a `session` level fixture isn't really a `session` level fixture depending on where it's defined, and I think we could solve a _lot_of problems/confusion by leaning into that behavior, and having scopes work like relative scopes when they can't really be applying to the full scope the claim to be. For example, if I define a fixture inside a class, and give it a scope of `session`, it can't really be anything more than `class`, but I could see it being used that way to make sure it happens before `class` scoped fixtures for tests in that class. In line with this, in the setup plan output, these could be indented according to the scope their effectively limited to in order to make a cleaner, easier to follow setup plan.

Regarding the pytest-django stuff, this is the PR I made to solve it in response to the merging if the original take of this PR:

https://github.com/pytest-dev/pytest-django/pull/807

I'm on mobile, so hopefully the description there explains it well. 

All that said, I find myself losing the forest for the trees a bit, so I'm definitely gonna have to sleep on this, and come back with some fresh eyes and re-read everything haha

SalmonMode replied:
Ok, so I took a look back, and read through things a bit more carefully (although I still need a bit more caffeine haha). That said, I did spot an issue with how things currently work in regards to your work example, and this bit in the docs:

> autouse fixtures obey the scope= keyword-argument: if an autouse fixture has scope='session' it will only be run once, no matter where it is defined. scope='class' means it will be run once per class, etc.

Given these fixtures:

```py
@pytest.fixture(scope='session', autouse=True, params=[1, 2])
def s():
    pass

@pytest.fixture(scope='class', autouse=True)
def c():
    pass
```

One would expect `c` to only happen once and only once per class because it's autouse. The problem comes in when parameterization is involved (at least for a higher scope fixture).

Given these test classes, though:

```py
class TestX:
    def test_it(self):
        pass

class TestY:
    def test_it_again(self):
        pass
```

Because pytest has to execute every test once for the first param set of `s` before it can do it again for the second param set, that means it'll have to setup `c` twice for each class because pytest had to tear it down after each class for the first param set and those tests are still dependent on `c` because of autouse. In total, `c` would be executed 4 times.

<details>
<summary>Here's the setup plan (from pytest 6.1.2)</summary>

```
SETUP    S s[1]
      SETUP    C c
        test2.py::TestX::test_it[1] (fixtures used: c, s)
      TEARDOWN C c
test2.py::TestY::test_it_again[1]
      SETUP    C c
        test2.py::TestY::test_it_again[1] (fixtures used: c, s)
      TEARDOWN C c
test2.py::TestX::test_it[2]
TEARDOWN S s[1]
SETUP    S s[2]
      SETUP    C c
        test2.py::TestX::test_it[2] (fixtures used: c, s)
      TEARDOWN C c
test2.py::TestY::test_it_again[2]
      SETUP    C c
        test2.py::TestY::test_it_again[2] (fixtures used: c, s)
      TEARDOWN C c
TEARDOWN S s[2]
```

</details>

This means that your `package` scoped fixture for work could easily be run more than once per package if more than one package depends on it.

Unfortunately, because of parameterization, there's no way to guarantee an autouse fixture only happens once per its defined scope.

_NOW_, I have an alternate proposal that I _think_ might suit everyone's needs.

In @bluetech's example:

```py
import pytest

@pytest.fixture(scope="module")
def m():
    print('M SETUP')
    yield
    print('M TEARDOWN')

class TestIt:
    @classmethod
    @pytest.fixture(scope="class", autouse=True)
    def c(cls):
        print('C SETUP')
        yield
        print('C TEARDOWN')

    def test_f1(self):
        print('F1 CALL')

    def test_f2(self, m):
        print('F2 CALL')
```

<details>
<summary>setup plan</summary>

```
      SETUP    C c
        test3.py::TestIt::test_f1 (fixtures used: c)
    SETUP    M m
        test3.py::TestIt::test_f2 (fixtures used: c, m)
      TEARDOWN C c
    TEARDOWN M m
```
</details>

the `m` fixture, to me, happens out of place, because it doesn't respect the ordering of scopes, until teardown. This is also a problem for your scenario, where you definitely only want a fixture running once ever.

Ideally, I would like this to be the setup plan for that example:

```
    SETUP    M m
      SETUP    C c
        test3.py::TestIt::test_f1 (fixtures used: c)
        test3.py::TestIt::test_f2 (fixtures used: c, m)
      TEARDOWN C c
    TEARDOWN M m
```

But what if only `test_f1` is to be run? Then I would want this to be the setup plan:

```
      SETUP    C c
        test3.py::TestIt::test_f1 (fixtures used: c)
      TEARDOWN C c
```

Since the `m` fixture wasn't needed by anything, it could be cut out.

To me, this is _much_ easier to not just reason about, but also to orchestrate.

This has pretty large implications, and could be incredibly challenging to figure out how to do this not just effectively, but in a way that folks can leverage in a predictable way for more advanced usage. _But_, I have an idea, and I think you'll get quite a kick out of it.

**Fixture Resolution Order (FRO)**

It works _almost exactly_ like MRO (so it's a familiar concept), except the "child" we're trying to figure out the FRO of, is the master FRO. After the first pass, it can be sorted by scope, and this can be used to ensure stack-like behavior.

Keep in mind, that entries in the FRO aren't required to be in the stack. It'd be more of a reference for knowing what fixtures should be executed after others, and how to teardown to appropriate common points. It can also look at the tests that would actually run (so not just the ones that were collected) and considered their actual dependencies to find out the actual scopes that running certain fixtures would be necessary for, even if the first test in those scopes isn't actually dependent on them, so they can be preemptively run for those scopes while also not running for scopes that don't need them.

We can also apply some tweaks to it to make things operate more efficiently. For example, if autouse fixtures are shifted left in the FRO, then parameterized fixtures can be shifted as far right as can be (so as to affect as few other fixtures as possible), and if they're autouse parameterized fixtures, they can be shifted as right as possible among the autouse fixtures of their scope.

For your use case, since you only want `initialize_petsc` to run once ever, you would just need to change its scope to `session`, and since `set_system_locale` is parameterized, it would happen _after_ `initialize_petsc`.

Nothing would ever get executed that didn't need to be, parameterized fixtures would have as reduced a footprint as they reasonably could, stack-like behavior would be achieved, and I'm not sure, but I think this could also offer some opportunities for optimization in pytest's code.

What do you think?

nicoddemus replied:
> To clarify, I'm not saying that the autouse fixture has to be dependent on the higher level fixture. I'm saying that if there's a need for (in the example), s[1] to happen before p, then p must be dependent on s[1], but if it isn't, then it can happen after p, and if it can happen after p, then it can be restructured slightly so that s happens after p and each param for s could apply to a separate version of that test without p having to be run again.

IIUC what you mean, then I agree and it is what I've been trying to convey: if a certain fixture requires the work from another fixture to happen before it, then it should explicitly declare a dependency on it. If `p` needs `s` in some way (due to a side effect like DB initialization, or because it needs the actual value produced by `s`) then `p` should depend on `s` (IOW, declare `s` as a paramater).

I took a look at pytest-dev/pytest-django#807 (thanks for the link), and it seems what happened there was exactly that: the `django_db_blocker` implicitly depended on the database (meaning: it needed the database to be setup at that point, but did not received any other fixture as parameter), which broke once a bug in pytest was fixed. The solution to the problem was to _explicitly_ make `django_db_blocker` depend on `django_db_setup`, which is also what I advocate here is correct and how things should work.

After reading back a few messages, I think an interpretation (which is also yours) is a bit incorrect and leads to some conclusions which are not true:

* The _scope_ of a fixture dictates when it is "active", our that it should execute during that scope. So a session-scoped fixture should be active during the entire session, and a class-scoped fixture will be active while the tests of a class execute.

When in fact what it means in pytest (currently implemented) is:

* The fixture scope dictates when a fixture will be _torn down_. A fixture will be active only when requested, and it will be teardown based on its scope: a session scope fixture will be executed when first requested, and torn down at the end of the session. A class-scoped fixture will be executed when first requested, and torn down after the last test of the class finishes.

"be requested" here also imply being requested by other fixtures, not only tests.

> One would expect c to only happen once and only once per class because it's autouse. The problem comes in when parameterization is involved (at least for a higher scope fixture).
> Unfortunately, because of parameterization, there's no way to guarantee an autouse fixture only happens once per its defined scope.

You are right; once parametrization plays in, the weak guarantee that a fixture at a high-level scope is executed once per scope breaks.

> the m fixture, to me, happens out of place, because it doesn't respect the ordering of scopes, until teardown. 

Indeed, it follows the rule I described above... but given `m` and  `c` don't depend on each other, I actually don't think it is a problem, because `m` apparently doesn't care about which order it will be torn down related to `c`. A bit weird I agree, but I don't think `m` should be setup before `test_f2`, as it is only requested then.

> Fixture Resolution Order (FRO)
> Nothing would ever get executed that didn't need to be, parameterized fixtures would have as reduced a footprint as they reasonably could, stack-like behavior would be achieved, reasoning about and planning setup plans would be easier (IMO), and I'm not sure, but I think this could also offer some opportunities for optimization in pytest's code.

Yeah I think you are onto something, indeed a clear view or different approach on how pytest deals with fixtures is something we have been wanting to tackle in a while: https://github.com/pytest-dev/pytest/issues/4871.

Currently the fixture mechanism is implemented as:

1. Lazy: fixtures are created only when first requested by a test or other fixture.
2. Implicit finalizers: the finalization stack is implicit, we don't have an eagle's eye view of which fixtures depend on which fixtures; it all happens by a series of finalizer callbacks being registered on each other to ensure things are torn down in the proper order.

There's definitely room for improvement there, however I do think we should keep 1) working as is.

Perhaps we should move the discussion there and close this PR, or do you think we should analyse this some more? Also I didn't hear from others in while here (understandable, keeping up with the discussion here takes a lot of time!).

Again, thanks for the rich discussion and patience here @SalmonMode!

SalmonMode replied:
Haha yes, I agree. Getting here using the GitHub mobile app is also pricing challenging (I wanted to give it a try, but it definitely needs some work).

And I agree. Given the nuances of this, the needs of pytest itself, its needs going forward, and backwards compatibility, I don't believe this PR is a sufficient fix. I've also started trying to implement the FRO approach to see how it could be done, and I can tell there's a lot of things that will need to be discussed. I'll move my proposal over to #4871, and this can be closed.

SalmonMode commented:
No worries!

bluetech commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
+            current_fix_index = len(request.fixturenames)
+        parent_fixture_indexes = set()
+
+        parent_request = getattr(request, "_parent_request")
```
```suggestion
        parent_request = request._parent_request
```

bluetech commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
+
+        stack_slice_index = min([current_fix_index, *parent_fixture_indexes])
+        active_fixture_argnames = all_fix_names[:stack_slice_index]
+        return tuple(tuple(active_fixture_argnames) + self.argnames)
```
```suggestion
        return tuple(active_fixture_argnames) + self.argnames
```

bluetech commented on testing/python/fixtures.py:
```diff
@@ -4236,6 +4364,25 @@ def test_suffix(fix_combined):
+        b = request._get_active_fixturedef("b")
+        assert b._will_be_finalized_by_fixture(a)
+
+    def test_a_only_finishes_one(self, request):
```
Not sure about this one but I think you meant

```suggestion
    def test_a_only_finishes_once(self, request):
```

bluetech commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
+        return False
+
+    def _dependee_fixture_argnames(self, request: SubRequest) -> Tuple[str, ...]:
+        """A list of argnames for fixtures that this fixture depends on.
+        Given a request, this looks at the currently known list of fixture argnames, and
+        attempts to determine what slice of the list contains fixtures that it can know
+        should execute before it. This information is necessary so that this fixture can
+        know what fixtures to register its finalizer with to make sure that if they
+        would be torn down, they would tear down this fixture before themselves. It's
+        crucial for fixtures to be torn down in the inverse order that they were set up
+        in so that they don't try to clean up something that another fixture is still
+        depending on.
+        When autouse fixtures are involved, it can be tricky to figure out when fixtures
+        should be torn down. To solve this, this method leverages the ``fixturenames``
+        list provided by the ``request`` object, as this list is at least somewhat
+        sorted (in terms of the order fixtures are set up in) by the time this method is
+        reached. It's sorted enough that the starting point of fixtures that depend on
+        this one can be found using the ``self._parent_request`` stack.
+        If a request in the ``self._parent_request`` stack has a ``:class:FixtureDef``
+        associated with it, then that fixture is dependent on this one, so any fixture
+        names that appear in the list of fixture argnames that come after it can also be
+        ruled out. The argnames of all fixtures associated with a request in the
+        ``self._parent_request`` stack are found, and the lowest index argname is
+        considered the earliest point in the list of fixture argnames where everything
+        from that point onward can be considered to execute after this fixture.
+        Everything before this point can be considered fixtures that this fixture
+        depends on, and so this fixture should register its finalizer with all of them
+        to ensure that if any of them are to be torn down, they will tear this fixture
+        down first.
+        This is the first part of the list of fixture argnames that is returned. The last
+        part of the list is everything in ``self.argnames`` as those are explicit
+        dependees of this fixture, so this fixture should definitely register its
+        finalizer with them.
```
```suggestion
        """A list of argnames for fixtures that this fixture depends on.

        Given a request, this looks at the currently known list of fixture argnames, and
        attempts to determine what slice of the list contains fixtures that it can know
        should execute before it. This information is necessary so that this fixture can
        know what fixtures to register its finalizer with to make sure that if they
        would be torn down, they would tear down this fixture before themselves. It's
        crucial for fixtures to be torn down in the inverse order that they were set up
        in so that they don't try to clean up something that another fixture is still
        depending on.

        When autouse fixtures are involved, it can be tricky to figure out when fixtures
        should be torn down. To solve this, this method leverages the ``fixturenames``
        list provided by the ``request`` object, as this list is at least somewhat
        sorted (in terms of the order fixtures are set up in) by the time this method is
        reached. It's sorted enough that the starting point of fixtures that depend on
        this one can be found using the ``self._parent_request`` stack.

        If a request in the ``self._parent_request`` stack has a ``:class:FixtureDef``
        associated with it, then that fixture is dependent on this one, so any fixture
        names that appear in the list of fixture argnames that come after it can also be
        ruled out. The argnames of all fixtures associated with a request in the
        ``self._parent_request`` stack are found, and the lowest index argname is
        considered the earliest point in the list of fixture argnames where everything
        from that point onward can be considered to execute after this fixture.
        Everything before this point can be considered fixtures that this fixture
        depends on, and so this fixture should register its finalizer with all of them
        to ensure that if any of them are to be torn down, they will tear this fixture
        down first.

        This is the first part of the list of fixture argnames that is returned. The last
        part of the list is everything in ``self.argnames`` as those are explicit
        dependees of this fixture, so this fixture should definitely register its
        finalizer with them.
```

bluetech commented on testing/python/fixtures.py:
```diff
@@ -1725,6 +1725,134 @@ def test_world(self):
         reprec.assertoutcome(passed=3)
 
 
+class TestMultiLevelAutouseAndParameterization:
```
```suggestion
class TestMultiLevelAutouseAndParametrization:
```

bluetech commented on testing/python/fixtures.py:
```diff
@@ -1725,6 +1725,134 @@ def test_world(self):
 
+class TestMultiLevelAutouseAndParameterization:
+    def test_setup_and_teardown_order(self, testdir):
+        """Tests that parameterized fixtures effect subsequent fixtures. (#6436)
+        If a fixture uses a parameterized fixture, or, for any other reason, is executed
+        after the parameterized fixture in the fixture stack, then it should be affected
+        by the parameterization, and as a result, should be torn down before the
+        parameterized fixture, every time the parameterized fixture is torn down. This
+        should be the case even if autouse is involved and/or the linear order of
+        fixture execution isn't deterministic. In other words, before any fixture can be
+        torn down, every fixture that was executed after it must also be torn down.
```
```suggestion
        """Tests that parametrized fixtures affect subsequent fixtures. (#6436)
        
        If a fixture uses a parametrized fixture, or, for any other reason, is executed
        after the parametrized fixture in the fixture stack, then it should be affected
        by the parametrization, and as a result, should be torn down before the
        parametrized fixture, every time the parametrized fixture is torn down. This
        should be the case even if autouse is involved and/or the linear order of
        fixture execution isn't deterministic. In other words, before any fixture can be
        torn down, every fixture that was executed after it must also be torn down.
```

bluetech commented on testing/python/fixtures.py:
```diff
@@ -1725,6 +1725,134 @@ def test_world(self):
+        """
+        testdir.makepyfile(
+            test_auto="""
+            import pytest
+            def f(param):
+                return param
+            @pytest.fixture(scope="session", autouse=True)
+            def s_fix(request):
+                yield
+            @pytest.fixture(scope="package", params=["p1", "p2"], ids=f, autouse=True)
+            def p_fix(request):
+                yield
+            @pytest.fixture(scope="module", params=["m1", "m2"], ids=f, autouse=True)
+            def m_fix(request):
+                yield
+            @pytest.fixture(scope="class", autouse=True)
+            def another_c_fix(m_fix):
+                yield
+            @pytest.fixture(scope="class")
+            def c_fix():
+                yield
+            @pytest.fixture(scope="function", params=["f1", "f2"], ids=f, autouse=True)
+            def f_fix(request):
+                yield
+            class TestFixtures:
+                def test_a(self, c_fix):
+                    pass
+                def test_b(self, c_fix):
+                    pass
```
```suggestion
            import pytest

            def f(param):
                return param

            @pytest.fixture(scope="session", autouse=True)
            def s_fix(request):
                yield

            @pytest.fixture(scope="package", params=["p1", "p2"], ids=f, autouse=True)
            def p_fix(request):
                yield

            @pytest.fixture(scope="module", params=["m1", "m2"], ids=f, autouse=True)
            def m_fix(request):
                yield

            @pytest.fixture(scope="class", autouse=True)
            def another_c_fix(m_fix):
                yield

            @pytest.fixture(scope="class")
            def c_fix():
                yield

            @pytest.fixture(scope="function", params=["f1", "f2"], ids=f, autouse=True)
            def f_fix(request):
                yield

            class TestFixtures:
                def test_a(self, c_fix):
                    pass

                def test_b(self, c_fix):
                    pass
```

bluetech commented on testing/python/fixtures.py:
```diff
@@ -1725,6 +1725,134 @@ def test_world(self):
+        test_fixtures_used = (
+            "(fixtures used: another_c_fix, c_fix, f_fix, m_fix, p_fix, request, s_fix)"
+        )
+        expected = (
```
I recommend using `"""` strings with `textwrap.dedent` here, then the `\n`s can be removed.

bluetech commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
+    def _will_be_finalized_by_fixture(
+        self, fixturedef: Union["FixtureDef", PseudoFixtureDef]
+    ) -> bool:
+        """Whether or not this fixture be finalized by the passed fixture.
+        Every ``:class:FixtureDef`` keeps a list of all the finishers (tear downs) of
+        other ``:class:FixtureDef`` instances that it should run before running its own.
+        Finishers are added to this list not by this ``:class:FixtureDef``, but by the
+        other ``:class:FixtureDef`` instances. They tell this instance that it's
+        responsible for tearing them down before it tears itself down.
+        This method allows a ``:class:FixtureDef`` to check if it has already told
+        another ``:class:FixtureDef`` that the latter ``:class:FixtureDef`` is
+        responsible for tearing down this ``:class:FixtureDef``.
```
Note I changed `finishers` to `finalizers` here but maybe you meant to refer to the `finish()` method? It wasn't clear to me at first, so would be good to clarify either way.

SalmonMode replied:
Thanks! Definitely finalizers!

SalmonMode commented on changelog/6436.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+:class:`FixtureDef <_pytest.fixtures.FixtureDef>` objects now properly register their finalizers with autouse and
```
You got it! I'll change it

SalmonMode commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
         return result
 
+    def _will_be_finalized_by_fixture(
+        self, fixturedef: Union["FixtureDef", PseudoFixtureDef]
```
You would be correct. I'm updating it now. I was previously doing this because mypy complained, so I'm putting an explicit assert in so it knows not to expect a PseudoFixtureDef there

SalmonMode commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
+        another ``:class:FixtureDef`` that the latter ``:class:FixtureDef`` is
+        responsible for tearing down this ``:class:FixtureDef``.
+        """
+        for finalizer in getattr(fixturedef, "_finalizers", ()):
```
Yeah, that was the reason. But I checked it out and it doesn't look like a `PseudoFixtureDef` could even get there, so I just tapped into the attribute directly.

SalmonMode commented on src/_pytest/fixtures.py:
```diff
@@ -1052,6 +1054,77 @@ def execute(self, request: SubRequest) -> _FixtureValue:
+        responsible for tearing down this ``:class:FixtureDef``.
+        """
+        for finalizer in getattr(fixturedef, "_finalizers", ()):
+            if "request" in getattr(finalizer, "keywords", {}):
```
I agree that this isn't a super clean approach. But I think getting a proper type/mechanism in place could make this a much larger change. I'm not sure how you wanna proceed with it though.

SalmonMode commented:
Given our discussion [here](https://github.com/pytest-dev/pytest/pull/7511#discussion_r523413759), it's clear that, while there is an issue here, a solution for this problem is going to take a more nuanced solution so that it doesn't violate current expectations of how fixtures are intended to work. This change, as it stands, may solve the issue it presents, but would also potentially cause other issues that could break existing test suites that depend on those expectations.

As such, I will be closing this PR, and the discussion will be continued over at #4871.

nicoddemus commented:
Graet, thanks @SalmonMode!

RonnyPfannschmidt commented:
@SalmonMode I want to thank you for going not only to great lengths to understand the issues and working out potential solutions, but also having the fortitude to put it down when the devil in the details strikes
